An Introduction To Kubernetes (#90)
directus-labs/guest-authoring: a-short-introduction-to-kubernetes/index.md

---
title: 'Introduction to Kubernetes'
description: 'A short introduction to Kubernetes in a Directus context'
author:
  name: 'Mikke Schirén'
  avatar_file_name: 'mikke.png'
---

## Introduction

> 

Reviewer: Let's open with the problem. What happens without K8S and therefore why does it need to exist?

> Originated in Googles project Borg, which started in 2003, Kubernetes (K8s) is today the most popular open-source container orchestration system, and it automates Deployments, scaling and management of applications. And Kubernetes is in many ways, a complicated beast to handle, like a puzzle with many pieces. If you never used Kubernetes before, it is maybe not the easiest way forward for your self-hosted Directus project. But there a lot of gains with Kubernetes running an application like Directus.

Reviewer: ```suggestion
Kubernetes (K8s) is the most popular open-source container orchestration system, and it automates deployments, scaling, and management of applications. While Kubernetes is not the easiest way to run a self-hosted Directus project, there a lot of gains to be realized.
```

Just shortening it down for clarity. Also, love an Oxford comma.

> Good to remember that Kubernetes is not a traditional, all-inclusive PaaS (Platform as a Service) system, when you start up a new cluster, depending on which flavor your using, you need to install some, or all pieces yourself.

Reviewer: I'd suggest removing this - we haven't spoken about it in the context of a PaaS so I don't think we need to compare. And by introducing the problem it solves upfront, we make it easier to understand immediately.

> If you never used Kubernetes, you may be surprised that you very often already have it on your development computer. Kubernetes, in a simple flavor is for instance shipped with Docker Desktop for Mac. So the place to test out Kubernetes itself it is very often nearby. If you want to experiment with Kubernetes for the first time, I really recommend that you do so locally.

Reviewer: ```suggestion
If you never used Kubernetes, you may be surprised that you very often already have it on your development computer, so the place to test out Kubernetes itself it is very often nearby. If you want to experiment with Kubernetes for the first time, I really recommend that you do so locally.
```

> I set up my own Kubernetes cluster on Amazon some years ago - before they had their offering [EKS](https://aws.amazon.com/eks/) (Amazon Elastic Kubernetes Service). I learned a lot, but I would not recommend my worst enemy to do it. Today there are a lot of ready Kubernetes solutions for you to choose from - like EKS, [GKE](https://cloud.google.com/kubernetes-engine) (Google Kubernetes Engine), Microsoft [Azure Kubernetes Service](https://azure.microsoft.com/en-us/products/kubernetes-service) (AKS) etc. And if you want more control you can setup Kubernetes with [Rancher](https://www.rancher.com/) (which we mostly do these days).

Reviewer: Haha this is funny but as an educational piece we don't need to jump into your personal gripes with different vendors. 

```suggestion
There are a lot of hosted Kubernetes solutions for you to choose from - like Amazon's [EKS](https://aws.amazon.com/eks/), Google's [GKE](https://cloud.google.com/kubernetes-engine), and Microsoft's [AKS](https://azure.microsoft.com/en-us/products/kubernetes-service). 
```

Beginner post, so no need to mention Rancher.

I will do a walk through of some of the basic pieces of the Kubernetes puzzle, and what they mean in a Directus context.

> :::info Please note

Reviewer: ```suggestion
:::info Examples Only
```

The example files in this post, should not be used to create a Directus setup, they are just there to describe the different parts, please see the GitHub gist in the end of the post if you are interested in a full working example.

:::

> ## Yaml and API:s

Reviewer: ```suggestion
## YAML and API:s
```

Is `API:s` correct? Just checking.

All objects can be described and created with yaml, with a minimum requirement telling which api you are going to use, and what kind, like:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: directus
  labels:
    app: directus
spec:
  containers:
  - name: directus
    image: directus/directus:10.8.3
    ports:
      - name: http
        containerPort: 8055
        protocol: TCP
```

> ## Metadata and labels

Reviewer: ```suggestion
## Metadata and Labels
```

> Metadata is used to identify your Kubernetes objects, without them you soon will be lost, and even different parts of the Kubernetes puzzle requires them. Some metadata is a must, like the name - otherwise there is no identification. Labels and annotations adds metadata to your objects so you could easily find them or target them.

Reviewer: ```suggestion
Metadata is used to identify your Kubernetes objects, with different parts of the Kubernetes puzzle requiring them to be set. Some metadata is required, with labels and annotations allowing easier object identification when searching for or targeting them.
```

One of the reasons is then you create an object, like a Deployment, a random string is used for the pod it creates. So you to communicate with it, you need the metadata to identify it.

> 

Reviewer: Move pods section here as you just mention them

Another, extremely useful metadata is annotations, but that is out of the scope for this post.

## Containers

> Containers are the Docker images you deploy to your K8s cluster. There are mainly two kind of containers, initContainers and containers. initContainers runs before your containers start, like if you need to set permissions, or do some task like updating your Directus schema etc.

Reviewer: ```suggestion
Containers are the Docker images you deploy to your K8s cluster. There are mainly two kind of containers, `initContainers` and `containers`. `initContainers` runs before your containers start, like if you need to set permissions, or do some task like updating your Directus schema.
```

## Environment variables

> Environment variables could be added as part of your object, like the public url for Directus.

Reviewer: ```suggestion
Environment variables can be added as part of your object, like the URL of a Directus project.
```

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: directus-app
  labels:
    app: directus
spec:
  containers:
  - name: directus
    image: directus/directus:10.8.3
    ports:

>     - containerPort: 80

Reviewer: This should be indented?

Author: No.

    env:
      - name: PUBLIC_URL
        value: https://directus.app
```

You could also use Secrets and use with for setting environment variables, that I will not cover here.

## Pods

> A pod is the smallest deployable compute object in K8s, like the example above. A pod is what is running one or several docker containers, a pod could be Directus, MySQL or Redis. If your pod crashes, it's dead and doesn't restart - it's like when you start a docker container with `docker run directus` and suddenly something gets weird, or your resources run out. So you need something to handle pods with, and the logic they are started, with config and volumes, there is where Deployments and StatefulSets come in.

Reviewer: This reads great. Great examples, caveats, and lead-in to the next bit of knowledge.

## Deployments

A Deployment manages a set of Pods to run an application workload, usually one that doesn't maintain state (coming that later). A Deployment is a way to describe the pod(s) you want to run, if you need to mount volumes (for storing data) or config (like the config file for your Directus Deployment), also you add logic for resources is going to use (RAM, and CPU). So, with a Deployment, you create pods. Each Deployment, creates normally one pod (which, like state above, could have many containers).

> A Deployment could "update" your pod - and what happens is that K8s takes down the existing pod, and replaces with a new one, like when updating your Directus instance, or adding new config or env variables.

Reviewer: ```suggestion
A Deployment could "update" your pod - where K8s takes down the existing pod, and replaces with a new one, like when updating your Directus instance, or adding new config or environment variables.
```

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: directus-Deployment
  labels:
    app: directus
spec:
  selector:
    matchLabels:
      app: directus
  template:
    metadata:
      labels:
        app: directus
    spec:
      containers:
      - name: directus
        image: directus:1.32.3
        ports:
          - containerPort: 80
        env:
          - name: PUBLIC_URL
            value: https://directus.app
```

> ## StatefulSets

Reviewer: ```suggestion
## `StatefulSets`
```

Can you wrap all thesein backticks? Otherwise the spellchecker is gonna be upset.

> When Deployments normally doesn't should be used for a pod that use a state, like a database, you need something that should be used for that, here is when you should use StatefulSet. Where a Deployment could normally just replace a pod, a StatefulSet needs either several replicas, or need to be deleted before it is replaced. Deletion of a pod, doesn't mean you loose your data, as long you are saving that data to a persistent volume.

Reviewer: Can you rewrite this? It doesn't make sense so I can't review it 😅

```yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: directus-mariadb
  labels:
    app: mariadb
spec:
  replicas: 1
  selector:
    matchLabels:
      app: mariadb
  serviceName: directus-mariadb
  template:
    metadata:
      labels:
        app.kubernetes.io/name: mariadb
    spec:
      containers:
        - name: mariadb
          image: mariadb:10.9.7
          ports:
            - name: mariadb
              containerPort: 3306
          volumeMounts:
            - name: data
              mountPath: /bitnami/mariadb
            - name: config
              mountPath: /opt/bitnami/mariadb/conf/my.cnf
              subPath: my.cnf
      volumes:
        - name: config
          configMap:
            name: directus-mariadb
  volumeClaimTemplates:
    - metadata:
        name: data
        labels:
          app.kubernetes.io/instance: directus
          app.kubernetes.io/name: mariadb
      spec:
        accessModes:
          - "ReadWriteOnce"
        resources:
          requests:
            storage: "8Gi"
```

## ReplicaSets

> An application like Directus works with replicas if you set up it with external file storage, and you are not using SQLLite (there are still ways to have replicas with internal file storage and SQLite, but that is more advanced then I will cover in this post, and I don't recommend you do that).

Reviewer: ```suggestion
A replica is...

An application like Directus works with replicas if you set up it with external file storage, and you are not using SQLLite.
```

Reviewer: You need to explain what a replicia s

Here we are creating 3 replicas of Directus:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: directus-Deployment
  labels:
    app: directus
spec:
  replicas: 3
  selector:
    matchLabels:
      app: directus
  template:
    metadata:
      labels:
        app: directus
    spec:
      containers:
      - name: directus
        image: directus:1.32.3
        ports:
          - containerPort: 80
```

## Volumes and StorageClasses

> As you could see in the MariaDB StatefulSets above, we are mounting volumes. Volumes could be temporarily, like a temp dir, or they could be persistent and data stays between Deployments.

Reviewer: ```suggestion
As you could see in the MariaDB StatefulSets above, we are mounting volumes. Volumes could be temporarily, like a `temp` directory, or they could be persistent and data stays between Deployments.
```

Here is an example of a Deployment with an `emptyDir` (`/tmp`) - a non-persistent volume, when the pod restarts, it is empty again.

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: directus-Deployment
  labels:
    app: directus
spec:
  selector:
    matchLabels:
      app: directus
  template:
    metadata:
      labels:
        app: directus
    spec:
      containers:
      - name: directus
        image: directus/directus:10.8.3
        ports:
          - containerPort: 80
        env:
          - name: PUBLIC_URL
            value: https://directus.app
          volumeMounts:
            - name: tmp
              mountPath: /tmp
      volumes:
        - name: tmp
          emptyDir: {}
```

## ConfigMaps

> ConfigMaps are used to store different kind of config objects, it could be environment variables, or like a whole file, which later could be used to mount (like a volume), and the application uses that.

Reviewer: ```suggestion
ConfigMaps are used to store different kind of configuration objects like environment variables or a file-based configuration, which later could be used to mount (like a volume), for an application to use.
```

Like config for MariaDB:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: release-name-mariadb
  labels:
    app: mariadb
data:
  my.cnf: |-
    [mysqld]
    skip-name-resolve
    max_allowed_packet=16M
```

And then mounted:

```yaml
[...]
          volumeMounts:
            - name: config
              mountPath: /opt/bitnami/mariadb/conf/my.cnf
              subPath: my.cnf
      volumes:
        - name: config
          configMap:
            name: release-name-mariadb
[...]

> ```

Reviewer: Fix indentation

Author: spaces are her intentional, as this is further in in the file. should I keep them?

Reviewer: Oh I see. I have a better way to present these, but i'll handle it when publishing.

This means that you can mount your config to a docker container, so the container could be clean, without your projects config.

## Services

> Service are used to expose the running pods, so you could communicate between the pods or expose them in different ways. The docker port is not exposed by default, so you need a way to tell Directus where to find MariaDB as an example. There are different kind of Services, and here we only going to cover the most common one, `ClusterIP`. Like the name says the service gets an ip, an internal one, so communication could happen.

Reviewer: ```suggestion
Service are used to expose the running pods, so you could communicate between the pods or expose them in different ways. The docker port is not exposed by default, so you need a way to tell Directus where to find MariaDB as an example. There are different kind of Services, and here we only going to cover the most common one, `ClusterIP`. Like the name says the service gets an internal IP so communication can happen between...
```

Finish the sentence

Example service for the MariaDB StatefulSet could be:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: directus-mariadb
  labels:
    app: mariadb
  annotations:
spec:
  type: ClusterIP
  ports:
    - name: mysql
      port: 3306
      protocol: TCP
      targetPort: mysql
  selector:
    labels:
      app.kubernetes.io/name: mariadb

```

This finds the MariaDB StatefulSet (the selector part, looks for a label named `app.kubernetes.io/name`, and value `mariadb`), and exposes port 3306 with the name `directus-mariadb` (equals hostname) for the running pod.

> So in Directus, you could communicate with the MariaDB pod, setting the port and the hostname as environment variables in the Deployment, like:

Reviewer: ```suggestion
In Directus, you can communicate with the MariaDB pod, setting the port and the hostname as environment variables in the Deployment:
```

---
title: 'Introduction to Kubernetes'
description: 'A short introduction to Kubernetes in a Directus context'
author:
  name: 'Mikke Schirén'
  avatar_file_name: 'mikke.png'
---

## Introduction

> When you start to run applications in docker containers it is pretty simple just fire up a container. But then you need another container, like a database, and then you need secrets, hosts, backups, updates etc. It could be quite messy to handle all the things, and you need something to orchestrate it with.

Reviewer: ```suggestion
While running applications in Docker containers may not seem to complex, the reality of multi-container setups — like needing a database, secrets, hosts, backups, updates, and more — can get quite messy. This is where an orchestrator comes in.```
